optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,158 @@
+            `Asynchronous SuccessiveHalving paper <http://arxiv.org/abs/1810.05934>`_.
+            The minimum early stopping rate for ``i``th bracket is :math:`i + s`.
+    """
+
+    def __init__(
+            self,
+            min_resource=1,
+            reduction_factor=3,
+            n_brackets=4,
+            min_early_stopping_rate_low=0
+    ):
+        # type: (int, int, int, int) -> None
+
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_budget = 0
+        self._n_brackets = n_brackets
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hyperband has {} brackets'.format(self._n_brackets))
+
+        for i in range(n_brackets):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_brackets)
+            self._resource_budget += bracket_resource_budget
+            self._bracket_resource_budgets.append(bracket_resource_budget)
+
+            # N.B. (crcrpar): `min_early_stopping_rate` has the information of `bracket_index`.
+            min_early_stopping_rate = min_early_stopping_rate_low + i
+
+            _logger.debug(
+                '{}th bracket has minimum early stopping rate of {}'.format(
+                    i, min_early_stopping_rate))
+
+            pruner = SuccessiveHalvingPruner(
+                min_resource=min_resource,
+                reduction_factor=reduction_factor,
+                min_early_stopping_rate=min_early_stopping_rate,
+            )
+            self._pruners.append(pruner)
+
+    def prune(self, study, trial):
+        # type: (Study, structs.FrozenTrial) -> bool
+
+        i = self._get_bracket_id(study, trial)
+        _logger.debug('{}th bracket is selected'.format(i))
+        bracket_study = self._create_bracket_study(study, i)
+        return self._pruners[i].prune(bracket_study, trial)
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_brackets):
+        # type: (int, int) -> int
+
+        n = self._reduction_factor ** (n_brackets - 1)
+        budget = n
+        for i in range(pruner_index, n_brackets - 1):
+            budget += n / 2
+        return budget
+
+    def _get_bracket_id(self, study, trial):
+        # type: (Study, structs.FrozenTrial) -> int
+        """Computes the index of bracket for a trial of ``trial_number``.
+
+        The index of a bracket is noted as :math:`s` in
+        `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
+        """
+
+        n = hash('{}_{}'.format(study.study_name, trial.number)) % self._resource_budget
+        for i in range(self._n_brackets):
+            n -= self._bracket_resource_budgets[i]
+            if n < 0:
+                return i
+
+        assert False, 'This line should be unreachable.'
+
+    def _create_bracket_study(self, study, bracket_index):
+        # type: (Study, int) -> Study
+
+        from optuna.study import Study
```

Comment:
Please move this import to the top of this file to follow [PEP 8](https://www.python.org/dev/peps/pep-0008/#id23) (it says "Please move this import to the top of this file to align PEP 8").

Reply:
If I follow PEP8, the import will be circular. I'll document this as a comment.

Reply:
I see. Thank you for the explanation.

Reply:
You don't have to import `Study` here, or at all, if you import `optuna` at the top, as the class can be accessed through `optuna.study.Study`.

Reply:
Thank you for the suggestion.

I applied it as shown in [this change](https://gist.github.com/crcrpar/8666765efbf7ed92a363b281483478d9), but `pytest tests/pruners_tests/test_hyperband.py` failed with `AttributeError: module 'optuna' has no attribute 'study'`.

Reply:
I see, thank you for trying it out. The problem with moving out `_BracketStudy` to the submodule scope is that the base class `Study` will be resolved when `hyperband.py` is imported, and there's still a circular dependency involving `hyperband.py` (resolves `study`) and `study.py` (which resolved `pruners`) which causes your above error. 

A solution could be to move `_BracketStudy` to `study.py` but I'm not sure what others think about this. Maybe we can leave it out from this PR as it's a private class anyway.

Could you fix the import?

Reply:
> A solution could be to move `_BracketStudy` to `study.py`

I don't like this because currently only `HyperbandPruner` needs `_BracketStudy`.

Reply:
Import fixed!